packages/react/lib/common.js

```diff
@@ -1,25 +1,14 @@
 'use strict';
 
-exports.Context = function Context (options) {
-  this.options = options || {};
-  this.initialize(this.options);
-};
+var mixin = require('mixinable');
 
-exports.Context.extend = function (protoExtension, staticExtension) {
-  var parent = this;
-  var Context = function (options) {
-    if (!(this instanceof Context)) {
-      return new Context(options);
-    }
-    parent.call(this, options);
-  };
-  return Object.assign(Context, parent, staticExtension, {
-    prototype: Object.assign(
-      Object.create(parent.prototype),
-      protoExtension,
-      {
-        constructor: Context
-      }
-    )
-  });
-};
+exports.Context = mixin({
```

Comment:
The helper methods in your package are very useful, I think (although I'm sure, that there are packages outside, that can do the same thing - or native code, that is not much longer). Mixin itself on the other hand seems to be another class implementation, which landed in node at version 4.0.0 (http://node.green/#ES2015-functions-class).

Reply:
`class` is missing in IE, though, so we will not be able to use it in hops anytime soon. And I am still conviced that limiting ourselves to just one, pseudo-classical, form of inheritance in JS is not a good idea. #klassenkampf

What 3rd party implementations of those helpers are you thinking of? I would love to keep mixinable as small as possible...

Reply:
Oh ... then I have a misunderstanding of the flow. Isn't this transpiled by webpack for the browser?

And I'm not completely sure what you mean by 'And I am still conviced that limiting ourselves to just one, pseudo-classical, form of inheritance in JS is not a good idea.'... Should we make the API complicated to learn by intention?

Reply:
No, the hops stuff is not transpiled. What I am replacing with this this PR is the previous Backbone-style `extends` implementation. 

And no we should not complicate things arbitrarily, of course. But having `class`es in ES6+ does, in my opinion, not imply we have to stop using the actual underlying inheritance mechanism, i.e. prototypal inheritance. `class`es are "just" syntactic sugar on top of that tried and tested mechanism, after all.

In fact, I think that with this mechanism, we might actually simplify things for our users. This is what one would have to do to compose different contexts after this PR..:

```javascript
import { createContext } from 'hops-react';
import reduxMixin from 'hops-redux';
import graphqlMixin from 'hops-graphql';

export default createContext.mixin(
  reduxMixin,
  graphqlMixin
);
```

To achieve the same compositional pattern with `class`es, we would have to do things like..:

```javascript
import { Context as ReactContext } from 'hops-react';
import { Context as ReduxContext } from 'hops-redux';
import { Context as GraphQLContext } from 'hops-graphql';

export default class Context extends ReactContext {
  enhanceElement(reactElement) {
    return super.enhanceElement()
      .then(ReduxContext.prototype.enhanceElement.bind(this))
      .then(GraphQLContext.prototype.enhanceElement.bind(this));
  }
  getTemplateData() {
    return {
      ...super.getTemplateData(),
      ...ReduxContext.prototype.getTemplateData.call(this)
      ...GraphQLContext.prototype.getTemplateData.call(this)
    };
  }
}
```

Developing additional mixins is, in my opinion, rather similar to what one would do using `class`es. Am I making some sort of sense?

Reply:
I know what you want to accomplish, but the problem is, that `react` obviously always has to be the root in order to get this to work (as all packages currently depend on react and it is currently the only 'Mixinable' if I see correctly). If some other mixin depends on any other (like e.g. router on redux or dataEnhancer on redux), you'll get to the babel problem (every module is a dedicated plugin, but they all have to be **applied in the correct order**).

Second point: By using mixin, you can enhance an object however you want, e.g. with **additional non lifecycle functions**. Let's say mixin `fooDataEnhancer` implements a function `createEnhancer`. When using this together with hops-redux, it will (correct me if I'm wrong) not be added to the object and therefore not work if it's added after hops redux or kill hops redux when added before.

And also: Each of the lifecycle methods **can and does alter local state** of the context instance. So you can load something in `bootstrap` and later on rely on it in `enhanceElement`. I think it's probably better to be pure and uniform in a way, e.g. like making all lifecycle methods equally async and make them return data that can later on be picked up by the next lifecycle. 

__So isn't it better to go one step back and go simpler with plain objects?__

The object would contain keys of lifecycle hooks - super easily testable (even outside a lifecycle) as they only get input and deliver output with `then`.

The usage for an app would still be easy

```jsx
const myContext = [
  reactContext, // ?
  // reduxContext,
  graphQlContext,
  {
    bootstrap: function () {
      //do something funny
    }
    enhanceElement(el) {
      return Promise.resolve(<Enhancer>{el}</Enhancer>);
    }
  }
]

render(<MyRoot />, myContext)
```

Because an intelligent, central manager coordinates everything like
- stitching all together (i.e goes through all contexts and executes their bootstrap then goes to the next lifecycle)
- using the right method (parallel, serial or merge) to combine the promises
- handing over data to the next lifecycle hook if necessary
- calling stuff differently on the client and on the server
- ...

Reply:
Good points from where I'm standing - I'll take that approach for a spin in the next couple of days (or rather nights).